README.md
# sechm

**Pierre-Luc Germain, 08.04.2021**

*D-HEST Institute for Neurosciences, ETH Zürich & Laboratory of Statistical Bioinformatics, University Zürich*

***

The *sechm* package is essentially a convenient wrapper around the 
*[ComplexHeatmap](https://jokergoo.github.io/ComplexHeatmap-reference/book/)* 
engine, facilitating the creation of heatmaps from 
*[SummarizedExperiment](https://bioconductor.org/packages/3.10/SummarizedExperiment)*
objects. This was formerly part of the [SEtools](https://github.com/plger/SEtools) package.

Install with:
```r
BiocManager::install("plger/sechm")
```

Example usage:

```r
sechm(SE, features=myGenes, top_annotation="Time", do.scale=TRUE, gaps_at="Condition",
       mark=c("Gm14288","Fos","Arc"))
```

![](inst/docs/example.png)<!-- -->

See the package's [vignette](http://www.bioconductor.org/packages/release/bioc/vignettes/sechm/inst/doc/sechm.html) for more information!


inst/NEWS
Changes in version 1.9.4 (2023-07-14)
+ row ordering now changed in the heatmap rather than matrix
+ fixed a few bugs, added a few extras
+ enabled list of features as input
Changes in version 1.3.1 (2022-04-15)
+ fixed a few bugs
+ better support for default arguments
+ removed deprecated arguments
+ added meltSE
Changes in version 1.0.0 (2021-04-19)
+ sechm moved from the SEtools package
+ generalized annotations


R/sechm.R
#' sechm
#'
#' ComplexHeatmap wrapper for
#' \code{\link[SummarizedExperiment]{SummarizedExperiment-class}}.
#'
#' @param se A \code{\link[SummarizedExperiment]{SummarizedExperiment-class}}.
#' @param features A vector of features (i.e. row names of `se`). Alternatively,
#'   can be a list of feature sets, in which case these will be plotted as
#'   different row chunks.
#' @param do.scale Logical; whether to scale rows (default FALSE).
#' @param assayName An optional vector of assayNames to use. The first available
#'  will be used, or the first assay if NULL.
#' @param cluster_cols Whether to cluster columns (default F)
#' @param cluster_rows Whether to cluster rows; default FALSE if
#' `do.sortRows=TRUE`.
#' @param sortRowsOn Sort rows by MDS polar order using the specified columns
#' (default all)
#' @param toporder Optional vector of categories on which to supra-order when
#' sorting rows, or name of a `rowData` column to use for this purpose.
#' @param hmcols Colors for the heatmap.
#' @param breaks Breaks for the heatmap colors. Alternatively, symmetrical
#' breaks can be generated automatically by setting `breaks` to a numerical
#' value between 0 and 1. The value is passed as the `split.prop` argument to
#' the \code{\link{getBreaks}} function, and indicates the proportion of the
#' points to map to a linear scale, while the more extreme values will be
#' plotted on a quantile scale. `breaks=FALSE` will disable symmetrical scale
#' and quantile capping, while retaining automatic breaks. `breaks=1` will
#' produce a symmetrical scale without quantile capping.
#' @param gaps_at Columns of `colData` to use to establish gaps between columns.
#' @param gaps_row Passed to the heatmap function; if missing, will
#' be set automatically according to toporder.
#' @param left_annotation Columns of `rowData` to use for left annotation.
#' Alternatively, an `HeatmapAnnotation` object.
#' @param right_annotation Columns of `rowData` to use for left annotation.
#' Alternatively, an `HeatmapAnnotation` object.
#' @param top_annotation Columns of `colData` to use for top annotation.
#' Alternatively, an `HeatmapAnnotation` object. To disable (overriding
#'   defaults), use `top_annotation=character()`.
#' @param bottom_annotation Columns of `colData` to use for bottom annotation.
#' Alternatively, an `HeatmapAnnotation` object.
#' @param anno_colors List of colors to use for annotation.
#' @param annorow_title_side Side (top or bottom) of row annotation names
#' @param annocol_title_side Side (left or right) of column annotation names
#' @param name The name of the heatmap, eventually appearing as title of the
#' color scale.
#' @param show_rownames Whether to show row names (default TRUE if less than
#' 50 rows to plot).
#' @param show_colnames Whether to show column names (default FALSE).
#' @param show_annotation_legend Logical; whether to show the annotation legend.
#' @param includeMissing Logical; whether to include missing features (default
#' FALSE)
#' @param mark An optional vector of gene names to highlight.
#' @param isMult Logical; used to silence labels when plotting multiple heatmaps
#' @param show_heatmap_legend Logical; whether to show heatmap legend
#' @param sort.method Row sorting method (see \code{\link{sortRows}})
#' @param na_col Color of NA values
#' @param ... Further arguments passed to `Heatmap`
#'
#' @return A a \code{\link[ComplexHeatmap]{Heatmap-class}}.
#'
#' @examples
#' data("Chen2017", package="sechm")
#' sechm(Chen2017, row.names(Chen2017)[1:10], do.scale=TRUE)
#'
#' @importFrom circlize colorRamp2
#' @importFrom methods is
#' @import SummarizedExperiment
#' @import ComplexHeatmap
#' @export
#' @name sechm
#' @rdname sechm
sechm <- function(se, features, do.scale=FALSE, assayName=NULL,
                  name=NULL, sortRowsOn=seq_len(ncol(se)), cluster_cols=FALSE,
                  cluster_rows=is.null(sortRowsOn), toporder=NULL, hmcols=NULL,
                  breaks=.getDef("breaks"), gaps_at=NULL,
                  gaps_row=NULL, left_annotation=NULL,
                  right_annotation=NULL, top_annotation=NULL,
                  bottom_annotation=NULL, anno_colors=list(),
                  show_rownames=NULL, show_colnames=FALSE,
                  isMult=FALSE, show_heatmap_legend=!isMult,
                  show_annotation_legend=TRUE, mark=NULL, na_col="white",
                  annorow_title_side=ifelse(show_colnames,"bottom","top"),
                  annocol_title_side="right", includeMissing=FALSE,
                  sort.method="MDS_angle", ...){

  if(is.list(features)){
    if(!is.null(gaps_row))
      warning("`gaps_row` ignored when passing a list of features")
    gaps_row <- rep(factor(names(features)),lengths(features))
    names(gaps_row) <- features <- unlist(features)
  }
  stopifnot(is.vector(features))
  if(is.numeric(features) && all(features==round(features)) && all(features>=1))
      features <- row.names(se)[features]
  if(is.factor(features)) features <- as.character(features)
  stopifnot(is.character(features))

  assayName <- .chooseAssay(se, assayName, returnName = TRUE)
  if(is.null(name)){
      if(is.numeric(assayName)){
          name <- ifelse(do.scale, "z-scores", "assay")
      }else{
          name <- ifelse(do.scale, paste0("scaled\n",assayName), assayName)
      }
  }

  x <- .prepData(se, genes=features, do.scale=do.scale, assayName=assayName,
                 includeMissing=includeMissing )

  toporder <- .parseToporder(rowData(se)[row.names(x),,drop=FALSE], toporder)
  row_order <- NULL
  if(!is.null(sortRowsOn) && length(sortRowsOn)>0 && nrow(x)>2){
    x2 <- sortRows(x[,sortRowsOn,drop=FALSE], toporder=toporder,
                   na.rm=TRUE, method=sort.method)
    row_order <- row.names(x2)
  }

  if( is.null(breaks) ){
      if( (!is.null(assayName) && grepl("^log[2]?FC$|scaledLFC",assayName))
          || do.scale)  breaks <- 0.995
  }
  hmcols <- .getBaseHMcols(se, hmcols)
  cscale <- .prepScale(x, hmcols=hmcols, breaks=breaks)

  breaks <- cscale$breaks
  hmcols <- circlize::colorRamp2(breaks, cscale$hmcols)

  anno_colors <- .getAnnoCols(se, anno_colors)

  if(!is.null(mark) && length(mark <- which(row.names(x) %in% mark))>0){
    mark <- anno_mark(mark, row.names(x)[mark], which="row")
  }else{
    mark <- NULL
  }

  if(is.null(left_annotation)) left_annotation <- .defaultAnno(se, "left")
  if(is.null(right_annotation) && is.null(mark))
    right_annotation <- .defaultAnno(se, "right")
  if(is.null(top_annotation)) top_annotation <- .defaultAnno(se, "top")
  if(is.null(bottom_annotation)) bottom_annotation <- .defaultAnno(se, "bottom")

  if(!is(left_annotation,"HeatmapAnnotation")){
    if(length(left_annotation)>0 && is.character(left_annotation)){
      left_annotation <- .prepareAnnoDF(
        rowData(se)[row.names(x),,drop=FALSE], anno_colors,
        left_annotation, whichComplex="row", show_legend=show_annotation_legend,
        show_annotation_name=!is.na(annorow_title_side),
        anno_name_side=ifelse(is.na(annorow_title_side), "top",
                              annorow_title_side))
    }else{
      left_annotation <- NULL
    }
  }

  if(!is(right_annotation,"HeatmapAnnotation")){
    if(length(right_annotation)>0 && is.character(right_annotation)){
      right_annotation <- .prepareAnnoDF(
        rowData(se)[row.names(x),,drop=FALSE], anno_colors,
        right_annotation, whichComplex="row",
        show_legend=show_annotation_legend,
        show_annotation_name=!is.na(annorow_title_side), highlight=mark)
    }else{
      right_annotation <- NULL
    }
  }

  if(is.null(right_annotation) && !is.null(mark)){
    right_annotation <- rowAnnotation(highlight=mark)
  }

  if(!is(top_annotation,"HeatmapAnnotation")){
    if(length(top_annotation)>0 && is.character(top_annotation)){
      top_annotation <- .prepareAnnoDF(
        colData(se), anno_colors, top_annotation, whichComplex="column",
        show_legend=(show_annotation_legend && !isMult),
        show_annotation_name=(!is.na(annocol_title_side) && !isMult),
        anno_name_side=ifelse(is.na(annocol_title_side), "right",
                              annocol_title_side) )
    }else{
      top_annotation <- NULL
    }
  }
  if(!is(bottom_annotation,"HeatmapAnnotation")){
    if(length(bottom_annotation)>0 && is.character(bottom_annotation)){
      bottom_annotation <- .prepareAnnoDF(
        colData(se), anno_colors, bottom_annotation, whichComplex="column",
        show_legend=(show_annotation_legend && !isMult),
        show_annotation_name=(!is.na(annocol_title_side) && !isMult),
        anno_name_side=ifelse(is.na(annocol_title_side), "right",
                              annocol_title_side) )
    }else{
      bottom_annotation <- NULL
    }
  }

  if(is.null(gaps_at) && !is.null(metadata(se)$default_view))
    gaps_at <- metadata(se)$default_view$gridvar
  if(is.null(gaps_at)) gaps_at <- .getDef("gaps_at")
  if(!is.null(gaps_at) && length(gaps_at)!=ncol(se))
    gaps_at <- intersect(gaps_at, colnames(colData(se)))
  if(isFALSE(gaps_at)) gaps_at <- NULL
  gaps_col <- .getGaps(gaps_at, colData(se), silent=TRUE)
  gaps_row <- .getGaps(gaps_row, rowData(se)[row.names(x),,drop=FALSE])

  if(is.null(show_rownames)) show_rownames <- nrow(x)<50 && is.null(mark)
  if(nrow(x)<=2) cluster_rows <- FALSE

  Heatmap( x, col=hmcols, na_col=na_col, name=name, row_order=row_order,
    show_row_names=show_rownames, show_column_names=show_colnames,
    row_split=gaps_row, column_split=gaps_col, cluster_rows=cluster_rows,
    show_heatmap_legend=show_heatmap_legend, cluster_columns=cluster_cols,
    top_annotation=top_annotation, left_annotation=left_annotation,
    bottom_annotation=bottom_annotation, right_annotation=right_annotation,
    ...)
}


R/meltSE.R
#' meltSE
#'
#' Melts a SE object into a \code{\link[ggplot2]{ggplot}}-ready long data.frame.
#'
#' @param x An object of class
#' \code{\link[SummarizedExperiment]{SummarizedExperiment-class}}
#' @param features A vector of features (i.e. row.names) to include. Use
#'   `features=NULL` to include all.
#' @param assayName The name(s) of the assay(s) to use. If NULL and the assays
#'   are named, all of them will be included.
#' @param colDat.columns The colData columns to include (defaults includes all).
#'   Use `colDat.columns=NA` in order not to include any.
#' @param rowDat.columns The rowData columns to include (default all). Use
#'   `rowData=NA` to not include any.
#' @param flatten Logical, whether to flatten nested data.frames.
#' @param baseDF Logical, whether to return a base data.frame (removing columns
#'   containing other objects such as atomic lists). Filtering is applied after
#'   flattening.
#'
#' @return A data.frame (or a DataFrame).
#'
#' @examples
#' data("Chen2017", package="sechm")
#' head(meltSE(Chen2017,"Fos"))
#'
#' @import SummarizedExperiment
#' @importFrom S4Vectors DataFrame
#' @export
meltSE <- function(x, features, assayName=NULL, colDat.columns=NULL,
                   rowDat.columns=NULL, flatten=TRUE, baseDF=TRUE){
  features <- intersect(features, row.names(x))
  if(is.null(colDat.columns)) colDat.columns <- colnames(colData(x))
  if(all(is.na(colDat.columns))) colDat.columns <- c()
  colDat.columns <- intersect(colDat.columns, colnames(colData(x)))
  if(is.null(rowDat.columns)) rowDat.columns <- colnames(rowData(x))
  if(all(is.na(rowDat.columns))) rowDat.columns <- c()
  rowDat.columns <- intersect(rowDat.columns, colnames(rowData(x)))
  if(is.null(assayName) && !is.null(assayNames(x)) ) assayName <- assayNames(x)
  if(is.null(assayName)){
    a <- list(value=assay(x))
  }else{
    a <- assays(x)[assayName]
  }
  if(is.numeric(assayName)) names(a) <- paste0("assay", assayName)
  a <- lapply(a, FUN=function(x) x[features,,drop=FALSE])
  df <- data.frame( feature=rep(features,ncol(x)),
                    sample=rep(colnames(x), each=length(features)) )
  df <- cbind(df, .flattenDF(colData(x)[rep(seq_len(ncol(x)),each=length(features)),
                                        colDat.columns, drop=FALSE],
                             remove=!flatten))
  df <- cbind(df, .flattenDF(rowData(x)[rep(features, ncol(x)),
                                        rowDat.columns, drop=FALSE],
                             remove=!flatten))
  if(baseDF){
    for(f in colnames(df)){
      if(!is.vector(df[[f]]) && !is.factor(df[[f]])) df[[f]] <- NULL
    }
    df <- as.data.frame(df)
  }
  for(v in names(a)) df[[v]] <- as.numeric(a[[v]])
  df
}


.flattenDF <- function(x, columns=colnames(x), remove=FALSE){
  x <- DataFrame(x)[,columns,drop=FALSE]
  if(ncol(x)==0) return(x)
  isdf <- unlist(lapply(x, FUN=function(x) is.data.frame(x) || is(x, "DFrame")))
  for(f in names(x)[which(isdf)]){
    y <- x[[f]]
    colnames(y) <- paste(f, colnames(y), sep=".")
    x[[f]] <- NULL
    if(!remove) x <- cbind(x, y)
  }
  if(!any(sapply(x, FUN=function(x) !is.vector(x) && !is.factor(x))))
    x <- as.data.frame(x)
  x
}


R/options.R
.options <- local({
  options <- list("assayName"=c("logFC", "log2FC", "logcpm", "lognorm"),
                  "anno_colors"=list(), hmcols=c("blue","black","yellow"),
                  "top_annotation"=c(
                    "Batch", "batch", "Condition","condition", "Group", "group",
                    "Dataset", "Genotype", "genotype", "cluster_id", "group_id",
                    "celltype"), gaps_at=c("Dataset","cluster_id"),
                  breaks=NULL)
  names(emptyAnnos) <- emptyAnnos <- c("bottom_annotation", "left_annotation",
                                       "right_annotation", "anno_rows",
                                       "anno_columns")
  options <- c(options, lapply(emptyAnnos, FUN=function(x) character(0)))

  env <- new.env(parent=emptyenv())

  list(set=function(variable, value) {
    stopifnot(is.character(variable), length(variable) == 1L, !is.na(variable))
    stopifnot(variable %in% names(options))
    env[[variable]] <- value
  }, get=function(variable, default=NULL) {
    if(variable %in% names(env)) return(env[[variable]])
    if(variable %in% names(options)) return(options[[variable]])
    warning("Unknown option ", variable)
    return(NULL)
  }, reset=function(){
    for(f in names(options)) env[[f]] <- options[[f]]
  })
})

.getDef <- function(x, ...) .options$get(x)

#' resetAllSechmOptions
#'
#' Resents all package options
#'
#' @return None
#'
#' @examples
#' resetAllSechmOptions()
#'
#' @export
resetAllSechmOptions <- function() .options$reset()

#' setSechmOption
#'
#' Sets a package-wide option for `sechm`
#'
#' @param variable The name of the variable to set
#' @param value The parameter value to save
#'
#' @return None
#'
#' @examples
#' setSechmOption("hmcols", value=c("blue","black","yellow"))
#'
#' @export
setSechmOption <- function(variable, value){
  .options$set(variable,value)
}

R/crossHm.R
#' crossHm
#'
#' Plot a multi-panel heatmap from a list of
#' \code{\link[SummarizedExperiment]{SummarizedExperiment-class}}.
#'
#' @param ses A (named) list of
#'  \code{\link[SummarizedExperiment]{SummarizedExperiment-class}} objects,
#'  with some matching row.names between them.
#' @param features A vector of features (i.e. row.names) to plot.
#' @param assayName The name of the assay to use; if multiple names are given,
#' the first available will be used. Defaults to "logcpm", "lognorm".
#' @param only.common Logical; whether to plot only rows common to all SEs
#' (default TRUE).
#' @param do.scale Logical; whether to scale rows in each SE (default TRUE).
#' @param uniqueScale Logical; whether to force the same colorscale for
#' each heatmap.
#' @param sortBy Names or indexes of `ses` to use for sorting rows (default all)
#' @param cluster_cols Logical; whether to cluster columns (default FALSE).
#' @param cluster_rows Logical; whether to cluster rows (default TRUE if
#' `do.sortRows=FALSE`, FALSE otherwise).
#' @param toporder Optional verctor of categories on which to supra-order when
#' sorting rows, or name of a `rowData` column to use for this purpose.
#' @param hmcols Colors for the heatmap.
#' @param breaks Breaks for the heatmap colors. Alternatively, symmetrical
#' breaks can be generated automatically by setting `breaks` to a numerical
#' value between 0 and 1. The value is passed as the `split.prop` argument to
#' the \code{\link{getBreaks}} function, and indicates the proportion of the
#' points to map to a linear scale, while the more extreme values will be
#' plotted on a quantile scale. `breaks=FALSE` will disable symmetrical scale
#' and quantile capping, while retaining automatic breaks. `breaks=1` will
#' produce a symmetrical scale without quantile capping.
#' @param gaps_at Columns of `colData` to use to establish gaps between columns.
#' @param gaps_row A named vector according to which rows will be split.
#' @param left_annotation Columns of `rowData` to use for left annotation.
#' @param top_annotation Columns of `colData` to use for top annotation.
#' @param name The title of the heatmap key.
#' @param anno_colors List of colors to use for annotation.
#' @param show_rownames Whether to show row names (default TRUE if 50 rows or
#' less).
#' @param show_colnames Whether to show column names (default FALSE).
#' @param rel.width Relative width of the heatmaps
#' @param merge_legends Logical; passed to
#' \code{\link[ComplexHeatmap]{draw-HeatmapList-method}}
#' @param ... Any other parameter passed to each call of
#' \code{\link[ComplexHeatmap]{Heatmap}}.
#'
#' @return A Heatmap list.
#'
#' @examples
#' data("Chen2017", package="sechm")
#' se1 <- Chen2017[,1:6]
#' se2 <- Chen2017[,7:15]
#' se3 <- crossHm(list(se1=se1, se2=se2), row.names(se1)[1:10] )
#'
#' @importFrom circlize colorRamp2
#' @importFrom methods is
#' @importFrom S4Vectors metadata metadata<- SimpleList
#' @import SummarizedExperiment
#' @import ComplexHeatmap
#' @export
crossHm <- function(ses, features, do.scale=TRUE, uniqueScale=FALSE,
                    assayName=.getDef("assayName"), sortBy=seq_along(ses),
                    only.common=TRUE, cluster_cols=FALSE,
                    cluster_rows=is.null(sortBy), toporder=NULL, hmcols=NULL,
                    breaks=.getDef("breaks"), gaps_at=.getDef("gaps_at"),
                    gaps_row=NULL,  name=NULL,
                    top_annotation=.getDef("anno_columns"), 
                    left_annotation=.getDef("anno_rows"), 
                    anno_colors=list(), show_rownames=NULL, merge_legends=FALSE,
                    show_colnames=FALSE, rel.width=NULL, ... ){

  stopifnot(is.vector(features))
  if(is.factor(features)) features <- as.character(features)
  stopifnot(is.character(features))
  
  if(is(ses,"SummarizedExperiment")) ses <- list(ses)
  if(is.null(names(ses))) names(ses) <- paste("SE", seq_along(ses))
  if(!is.null(rel.width) && length(rel.width)!=length(ses))
      stop("If given, `rel.width` should have the same length as `ses`.")
  if(is.null(rel.width)) rel.width <- rep(1,length(ses))

  tt <- table(unlist(lapply(ses,row.names)))
  if(only.common){
      features <- intersect(features, names(tt)[which(tt==length(ses))])
  }else{
    features <- intersect(features, names(tt))
  }
  if(length(features)==0)
      stop("There appears to be not feature in common across `ses`")
  if(length(features)<=2){
    sortBy <- NULL
    cluster_rows <- FALSE
  }
  if(!is.null(toporder)){
      if(is.null(names(toporder)))
          stop("`toporder` should be a vector named by feature")
      if(!all(features %in% names(toporder)))
          warning("Some features are missing from `toporder`")
      toporder <- toporder[features]
      names(toporder) <- features
  }

  dats <- lapply( ses, FUN=.prepData, genes=features, assayName=assayName,
                  do.scale=do.scale && !uniqueScale, includeMissing=TRUE )
  if(do.scale && uniqueScale){
      x <- do.call(cbind, dats)
      x <- safescale(x, byRow=TRUE)
      dl <- lapply(dats, FUN=function(x) seq_len(ncol(x)))
      dl2 <- c(0,cumsum(lengths(dl)[-length(dl)]))
      dats <- lapply( seq_along(dl), FUN=function(i)
          x[,dl[[i]]+dl2[[i]],drop=FALSE] )
      names(dats) <- names(dl)
  }else{
      x <- NULL
  }

  if(uniqueScale){
      if(is.null(x)) x <- do.call(cbind, dats)
      if(is.null(breaks) && do.scale) breaks <- 0.995
      cscale <- .prepScale(x, hmcols=.getHMcols(hmcols), breaks=breaks)
      breaks <- cscale$breaks
      hmcols <- cscale$hmcols
  }

  if(!is.null(sortBy) && length(sortBy)>0){
      xs <- dats
      if(do.scale && !uniqueScale)
          xs <- lapply(xs, byRow=TRUE, FUN=safescale)
      xs <- do.call(cbind, xs[sortBy])
      features <- row.names(sortRows(xs,toporder=toporder))
      dats <- lapply(dats, FUN=function(x) x[features,,drop=FALSE])
  }

  CDs <- lapply(ses, ac=top_annotation, FUN=function(x,ac){
      ac <- intersect(colnames(colData(x)),ac)
      as.data.frame(colData(x)[,ac,drop=FALSE])
  })

  # make sure factors share the levels across datasets
  facts <- unique(unlist(lapply(CDs, FUN=function(x){
      x <- vapply(x, class, character(1))
      names(x)[x=="factor"]
  })))
  for(v in facts){
      lvls <- unique(unlist(lapply(CDs, FUN=function(x){
          if(v %in% colnames(x)) return(levels(droplevels(as.factor(x[[v]]))))
          NULL
      })))
      CDs <- lapply(CDs, FUN=function(x){
          if(!(v %in% colnames(x))) return(x)
          x[[v]] <- factor(as.character(x[[v]]), levels=lvls)
          x
      })
  }

  ses <- lapply(seq_along(ses), FUN=function(i){
      se <- tryCatch({
        RD <- as.data.frame(rowData(ses[[i]])[features,,drop=FALSE])
        row.names(RD) <- features
        SummarizedExperiment( list(a=dats[[i]]), colData=CDs[[i]],
                              rowData=RD, metadata=metadata(ses[[i]]) )
        }, error=function(e){
                       SummarizedExperiment(list(a=dats[[i]]), colData=CDs[[i]],
                                            metadata=metadata(ses[[i]]) )
        })
      row.names(se) <- features
      se
  })
  names(ses) <- names(dats)
  tmp <- unlist(lapply(ses, FUN=function(x) colnames(rowData(x))))
  afields <- intersect(c(top_annotation, left_annotation),
                       c(tmp, colnames(CDs[[1]])))
  anno_colors <- .getAnnoCols(ses[[1]], anno_colors, do.assign=TRUE)
  anno_colors <- anno_colors[afields]

  if(is.null(show_rownames)) show_rownames <- length(features)<50

  hlp <- list()
  if(uniqueScale){
      if(!is.null(assayName) && length(assayName)==1 &&
         !is.numeric(assayName)){
          hlp$title <- ifelse(do.scale, paste0("scaled\n",assayName),
                              assayName)
      }else{
          hlp$title <- ifelse(do.scale, "z-scores", "")
      }
  }

  htlist <- lapply(seq_along(ses), FUN=function(i){
      sechm(ses[[i]], features=features, do.scale=(do.scale && !uniqueScale),
            assayName="a", name=names(ses)[i], toporder=toporder,
            hmcols=hmcols, breaks=breaks, left_annotation=left_annotation,
            top_annotation=top_annotation, anno_colors=anno_colors,
            cluster_rows=cluster_rows, cluster_cols=cluster_cols,
            show_rownames=(show_rownames && i==length(ses)), sortRowsOn=NULL,
            show_colnames=show_colnames, isMult=i!=length(ses),
            show_heatmap_legend=(!uniqueScale || i==length(ses)),
            heatmap_legend_param=hlp, column_title=names(ses)[i],
            show_annotation_legend=FALSE, includeMissing=!only.common,
            width=rel.width[i], ...)
  })

  ht <- NULL
  for(f in htlist) ht <- ht + f
  if(length(anno_colors)>0)
      return(ComplexHeatmap::draw(ht,
              annotation_legend_list=.annoLegend(anno_colors),
              merge_legends=merge_legends))
  ht
}
